marlowe-spec-test/src/Marlowe/Spec/ClientProcess.hs

```diff
@@ -69,57 +93,108 @@ instance IsOption CliPath where
   optionCLParser = mkOptionCLParser (short 'c' <> metavar "CLI_PATH")
 
 
-newtype TraceCP = TraceCP Bool
+newtype TraceCP = TraceCP (Maybe FilePath)
   deriving (Show, Eq, Typeable)
 
-
 instance IsOption TraceCP where
-  defaultValue = TraceCP False
-  parseValue str = TraceCP <$> safeReadBool str
+  defaultValue = TraceCP Nothing
+  parseValue str = Just $ TraceCP $ Just str
   optionName = return "trace-cp"
-  optionHelp = return "Trace client process"
-  optionCLParser = mkFlagCLParser mempty (TraceCP True)
+  optionHelp = return "If provided, filepath of the client process trace log"
+  optionCLParser = mkOptionCLParser (metavar "TRACE_CP")
+
+newtype PoolSize = PoolSize Int
+  deriving (Show, Eq, Typeable)
+
+instance IsOption PoolSize where
+  defaultValue = PoolSize 10
+  parseValue str = PoolSize <$> safeRead str
+  optionName = return "pool-size"
+  optionHelp = return "Number of client process to spawn (default 10)"
+  optionCLParser = mkOptionCLParser (metavar "POOL_SIZE")
+
+withThreadProcess :: ClientProcessPool -> (ClientProcess -> IO a) -> IO a
+withThreadProcess clientProcess callback = do
+  let
+    pool = threadPool clientProcess
+  threadProcess <- atomically $ readTQueue pool
+
+  res <- callback threadProcess
+  atomically $ writeTQueue pool threadProcess
 
-withClientProcess :: (IO ClientProcess -> TestTree) -> TestTree
+  pure res
+
+debug :: Maybe Handle -> String -> IO ()
+debug mh str = void $ forM mh (\h -> hPutStrLn h $ C.pack str)
+
+createThreadProcess :: Int -> FilePath -> Maybe Handle -> IO ClientProcess
+createThreadProcess pid cliPath debugH = do
+  debug debugH $ show pid ++ " - Create client process with path: " ++ show cliPath
+
+  (Just hIn, Just hOut, _, ph) <- createProcess
+            (proc cliPath [])
+            { std_in  = CreatePipe
+            , std_out = CreatePipe
+            }
+  hSetBuffering hIn  LineBuffering
+  hSetBuffering hOut LineBuffering
+  pure
+      ClientProcess
+        { stdInH  = hIn
+        , stdOutH = hOut
+        , procH   = ph
+        , cpId    = pid
+        , traceH  = debugH
+        }
+
+withClientProcess :: (IO ClientProcessPool -> TestTree) -> TestTree
 withClientProcess testCreator =
   askOption
     (\cliPathOpt ->
       askOption
-        (\traceOpt -> withClientProcess' traceOpt cliPathOpt
+        (\traceOpt ->
+          askOption
+            (\poolSize ->
+              withClientProcess' traceOpt cliPathOpt poolSize
```

Comment:
Use an eta reduction here, instead?

Reply:
Two reductions gives:
```haskell
withClientProcess testCreator =
  askOption
    (\cliPathOpt ->
      askOption
        $ askOption
        . flip withClientProcess' cliPathOpt
```
and three reductions gives
```haskell
withClientProcess testCreator =
  askOption
    $ askOption
    . (askOption .)
    . flip withClientProcess'
```

It depends how pointfree you want to make this.